lms/djangoapps/branding_by_referer/middleware.py

```diff
@@ -0,0 +1,165 @@
+        """
+        Method to override the mktg links of the current site configuration
+        """
+        mktg_urls = configuration_helpers.get_value("MKTG_URLS")
+        if not mktg_urls:
+            return
+
+        for name, url in iteritems(mktg_urls):
+            url_domain = urlparse(url).netloc
+            if not url_domain:
+                continue
+            mktg_urls[name] = url.replace(url_domain, referer_domain)
+
+    def _get_referer_configurations(self, referer_domain):
+        """
+        This method extracts the override configurations for a specific referer
+        """
+
+        referers_configurations = configuration_helpers.get_value(self.CONFIGURATION_BY_REFERER_KEY)
+
+        if not referers_configurations:
+            return None
+
+        return referers_configurations.get(
+            referer_domain
+        )
+
+    def _insert_referer_configurations(self, referer_domain):
+        """
+        This method merges the override configurations of a specific referer with the current site configuration
+        """
+        current_referer_configuration = self._get_referer_configurations(referer_domain)
+        if not current_referer_configuration:
+            return
+
+        current_conf_values = configuration_helpers.get_current_site_configuration().values
+        for key, value in iteritems(current_referer_configuration):
+            if isinstance(value, dict):
+                try:
+                    merged = current_conf_values.get(key, {}).copy()
+                except AttributeError:
+                    merged = {}
+                merged.update(value)
+                current_conf_values[key] = merged
+                continue
+            current_conf_values[key] = value
+
+        return
+
+    def _update_conf(self, request, referer_domain):
+        """
+        Update the current site configuration based on the referer
+        """
+        # Adding a key to the current configuration to identify it as a overriden set
+        current_conf = configuration_helpers.get_current_site_configuration()
+        current_conf.values[self.OVERRIDE_MKTG_REFERER_KEY] = referer_domain
+
+        self._override_marketing_urls(referer_domain)
+        self._insert_referer_configurations(referer_domain)
+        return
+
+    def get_stored_referer_preference(self, request):
+        """
+        Read the referer domain value from the current logged-in user preferences
+        """
+
+        if not request.user.is_authenticated():
+            return None
+
+        stored_referer_data = UserPreference.get_value(request.user, self.MARKETING_SITE_REFERER)
+        if not stored_referer_data:
+            return None
+
+        try:
+            stored_referer_data_json = json.loads(stored_referer_data)
+            referer_domain_preference = stored_referer_data_json['referer_domain']
+        except ValueError:
+            # TODO: log something
```

Comment:
Since there are some UserPreference that doesn't have a json value, just a string, you must make sure to convert them into json values, and save them as well. @jfavellar90

Reply:
@Squirrel18 the BrandingByReferer middleware fixes that (I think it should)? if it does, I should rely only on the json value (as this new middleware will be located below BrandingByReferer), if not, we should have a helper method as a single source of the truth which can be used anywhere.

Reply:
@Squirrel18 in fact BrandingByReferer does the job as I see in the code

Reply:
Right, i thought that this middleware ran first. @jfavellar90